src/main/java/com/bootcamp/demo/service/BookingService.java

```diff
@@ -0,0 +1,170 @@
+        LinkedHashSet<Booking> allBookings;
+        return allBookings = db.collection(COLLECTION_PATH)
+                .get()
+                .get()
+                .getDocuments()
+                .stream().map(d -> d.toObject(Booking.class))
+                .collect(Collectors.toSet())
+                .stream().sorted(Comparator.comparing(Booking::getStartDate))
+                .collect(Collectors.toCollection(LinkedHashSet::new));
+    }
+
+    /**
+     * Creates a new document containing the given Booking type param.
+     */
+    @Override
+    public String createBooking(final Booking booking) throws ExecutionException, InterruptedException, IllegalArgumentException {
+        if (booking == null) throw new IllegalArgumentException("Parameter 'booking' cannot be null");
+        return db.collection(COLLECTION_PATH)
+                .document(booking.getId())
+                .set(booking)
+                .get()
+                .getUpdateTime()
+                .toString();
+    }
+
+    @Override
+    public String deleteBooking(final String bookingId) throws ExecutionException, InterruptedException {
+        if (bookingId == null) throw new IllegalArgumentException("Parameter 'bookingId' cannot be null");
+        return db.collection(COLLECTION_PATH)
+                .document(bookingId)
+                .delete()
+                .get()
+                .getUpdateTime()
+                .toString();
+    }
+
+    @Override
+    public String updateBooking(String bookingId, String endDate, PaymentStatus newStatus) throws ExecutionException, InterruptedException {
+        if (bookingId == null) throw new IllegalArgumentException("Parameter 'bookingId' cannot be null");
+        return db.collection(COLLECTION_PATH)
+                .document(bookingId)
+                .update("endDate", endDate, "paymentStatus", newStatus)
+                .get()
+                .getUpdateTime()
+                .toString();
+    }
+
+    public Scooter bookAScooter(BookScooter bookScooter) throws ExecutionException, InterruptedException {
+        List<Scooter> allAvailableScooters = scooterService.getAvailableScooters(
+                new Location(bookScooter.getLatitude(), bookScooter.getLongitude()), bookScooter.getSearchRadius());
+        return  ! allAvailableScooters.isEmpty()
+                ? bookRandomScooter(allAvailableScooters, bookScooter)
+                : null;
+    }
+
+    private Scooter bookRandomScooter(List<Scooter> allAvailableScooters, BookScooter bookScooter) throws ExecutionException, InterruptedException {
+        final Scooter pickedScooter = reserveScooter(allAvailableScooters);
+        createBookingScooter(pickedScooter, bookScooter);
+        return pickedScooter;
+    }
+
+    private void createBookingScooter(Scooter scooter, BookScooter bookScooter) throws ExecutionException, InterruptedException {
+        Booking booking = new Booking();
+        booking.setId(bookScooter.getId());

```

Comment:
Why is there an id here? Shouldn't the booking be created now? So as such, shouldn't it be a generated id rather than a provided one?

Reply:
That's right, my bad